File: nim/tokens.nim
import std/[strutils, options]

type
  Name* = enum
    CP, MV, RM, MKDIR, CHECK_ENV, SET_ENV, UNSET_ENV, EXEC, PUTS, PROMPT,
    CONFIRM, IF, THEN, ELSE, END, COND, FOR, DO, LABEL, GOTO, DOWNLOAD, UPLOAD,
    UNKNOWN

  Token* = object
    name: Name
    args: Option[seq[string]]

proc strToName*(name: string): Name =
  case name
  of "cp": return Name.CP
  of "mv": return Name.MV
  of "rm": return Name.RM
  of "mkdir": return Name.MKDIR
  of "check_env": return Name.CHECK_ENV
  of "set_env": return Name.SET_ENV
  of "unset_env": return Name.UNSET_ENV
  of "exec": return Name.EXEC
  of "puts": return Name.PUTS
  of "prompt": return Name.PROMPT
  of "confirm": return Name.CONFIRM
  of "if_": return Name.IF
  of "then": return Name.THEN
  of "else_": return Name.ELSE
  of "end": return Name.END
  of "cond": return Name.COND
  of "for_": return Name.FOR
  of "do": return Name.DO
  of "label": return Name.LABEL
  of "goto": return Name.GOTO
  of "download": return Name.DOWNLOAD
  of "upload": return Name.UPLOAD
  else: return Name.UNKNOWN

proc nameToStr*(name: Name): string =
  case name
  of Name.CP: "cp"
  of Name.MV: "mv"
  of Name.RM: "rm"
  of Name.MKDIR: "mkdir"
  of Name.CHECK_ENV: "check_env"
  of Name.SET_ENV: "set_env"
  of Name.UNSET_ENV: "unset_env"
  of Name.EXEC: "exec"
  of Name.PUTS: "puts"
  of Name.PROMPT: "prompt"
  of Name.CONFIRM: "confirm"
  of Name.IF: "if"
  of Name.THEN: "then"
  of Name.ELSE: "else"
  of Name.END: "end"
  of Name.COND: "cond"
  of Name.FOR: "for"
  of Name.DO: "do"
  of Name.LABEL: "label"
  of Name.GOTO: "goto"
  of Name.DOWNLOAD: "download"
  of Name.UPLOAD: "upload"
  of Name.UNKNOWN: "unknown"

proc getName*(token: Token): Name =
  token.name

proc getNameStr*(token: Token): string =
  nameToStr(token.name)

proc getArgs*(token: Token): seq[string] =
  if token.args.isSome:
    token.args.get()
  else:
    @[] 

proc initToken*(name: string, line: string): Token =
  let splitArgs = line.split(' ')
  result = Token(name: strToName(name), args: some(splitArgs))

# may not be needed, since no allocation needed.
proc deinitToken*(token: var Token) =
  if token.args.isSome:
    token.args = none(seq[string])

File: cpp/tokens.cpp
#include "tokens.hpp"

Name Token::get_name() const
{
    return this->token.first;
}

std::string Token::get_name_str() const
{
    return name_to_str(this->token.first);
}

std::vector<std::string> Token::get_args() const
{
    return this->token.second;
}

Name str_to_name(std::string name)
{
    static const std::unordered_map<std::string, Name> name_map = {
        {"cp", Name::CP}, {"mv", Name::MV}, {"rm", Name::RM}, 
        {"mkdir", Name::MKDIR}, {"check_env", Name::CHECK_ENV}, 
        {"set_env", Name::SET_ENV}, {"unset_env", Name::UNSET_ENV}, 
        {"exec", Name::EXEC}, {"puts", Name::PUTS}, {"prompt", Name::PROMPT}, 
        {"confirm", Name::CONFIRM}, {"if", Name::IF}, {"then", Name::THEN}, 
        {"else", Name::ELSE}, {"end", Name::END}, {"cond", Name::COND}, 
        {"for", Name::FOR}, {"do", Name::DO}, {"label", Name::LABEL}, 
        {"goto", Name::GOTO}, {"download", Name::DOWNLOAD}, 
        {"variable", Name::VARIABLE}, {"upload", Name::UPLOAD}
    };
    auto it = name_map.find(name);
    return it != name_map.end() ? it->second : Name::UNKNOWN;
}

std::string name_to_str(Name name)
{
    std::string result;
    switch (name) {
        case Name::CP:         result = "cp";        break;
        case Name::MV:         result = "mv";        break;
        case Name::RM:         result = "rm";        break;
        case Name::MKDIR:      result = "mkdir";     break;
        case Name::CHECK_ENV:  result = "check_env"; break;
        case Name::SET_ENV:    result = "set_env";   break;
        case Name::UNSET_ENV:  result = "unset_env"; break;
        case Name::EXEC:       result = "exec";      break;
        case Name::PUTS:       result = "puts";      break;
        case Name::PROMPT:     result = "prompt";    break;
        case Name::CONFIRM:    result = "confirm";   break;
        case Name::IF:         result = "if";        break;
        case Name::THEN:       result = "then";      break;
        case Name::ELSE:       result = "else";      break;
        case Name::END:        result = "end";       break;
        case Name::COND:       result = "cond";      break;
        case Name::FOR:        result = "for";       break;
        case Name::DO:         result = "do";        break;
        case Name::LABEL:      result = "label";     break;
        case Name::GOTO:       result = "goto";      break;
        case Name::DOWNLOAD:   result = "download";  break;
        case Name::UPLOAD:     result = "upload";    break;
        case Name::VARIABLE:   result = "variable";  break;
        case Name::UNKNOWN:    result = "unknown";   break;
        default:               result = "ERROR";     break;
    }
    return result;
}